### PR TITLE
HCLIND-38 Add field source to turbonomics 

### PR DESCRIPTION
**Pull request by midir99**
### Description

Addressing [HCLIND-38](https://flexera.atlassian.net/browse/HCLIND-38):

Ensure that the generated incidents will be converted into spend recommendations, in the policy info block:

- set provider to a corresponding cloud provider
- set recommendation_type to rate reduction
- new field source set to Turbonomic
- set policy set to “Reserved Instances”

### Issues Resolved

It just adds a new field to the report delivered.

### Contribution Check List

- [x] All tests pass.
- [ ] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+# Parameters   #
+##################
+
+parameter "auth_cookie" do
+  type "string"
+  label "Authorization Cookie"
+  no_echo true
+  description "authorization cookie pulled from manual source"
+end
+
```
Comment (degarajesh) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:28]: I remember ask from Tatian, Please pass url as param, Apply the same for all the policies

File: cost/turbonomics/allocate_virtual_machines_recommendations/gcp/turbonomics_allocate_virtual_machines.pt [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+###############################################################################
+
+escalation "esc_email" do
+  automatic true
+  label "Send Email"
+  description "Send incident email"
+  email $param_email
+end
+
+###############################################################################
```
Comment (degarajesh) [inline cost/turbonomics/allocate_virtual_machines_recommendations/gcp/turbonomics_allocate_virtual_machines.pt:370]: remove cloud workflow comment

Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/gcp/turbonomics_allocate_virtual_machines.pt]: @degarajesh - Done!

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+###############################################################################
+
+escalation "esc_email" do
+  automatic true
+  label "Send Email"
+  description "Send incident email"
+  email $param_email
+end
+
+###############################################################################
```
[the remark this answers is not in the document]
Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt]: @degarajesh - Done!

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt (hunk 1 above)
Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:28]: By URL you mean this one?
image

Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:28]: @taterdog590 - Could you please confirm?

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Changelog
+
+## v2.1
+
+- Field **source** was added to the report.
```
Comment (taniaSol) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/CHANGELOG.md:5]: Maybe replace "report" with "template" or "info block"

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+- *Authorization Cookie"* - authorization cookie pulled from turbonomic login endpoint: [Credentials](https://xxxx.turbonomic.com/api/v3/login)
+  - no_echo: true
+- *Email addresses* - A list of email addresses to notify
+
+### Required Flexera Roles
+
+- policy_manager
+- billing_center_viewer
+
+- Turbonomics - administrator
```
Comment (taniaSol) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md:24]: Turbonomic administrator privileges are not required

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+##################
+# Parameters   #
+##################
+
+parameter "auth_cookie" do
+  type "string"
+  label "Authorization Cookie"
+  no_echo true
+  description "authorization cookie pulled from manual source"
```
Comment (taniaSol) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:26]: Perhaps change description to "Valid authorization cookie"

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt (hunk 1 above)
Comment (taniaSol) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:28]: Yes, this value should be passed as parameter

**Review by taniaSol — COMMENTED** [review]
Please replace "Turbonomics" with "Turbonomic" in user facing places

**Comment by midir99** [comment]
@taniaSol - I replaced "Turbonomics" with "Turbonomic" in user facing places 😁

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt (hunk 6 above)
Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt]: @taniaSol - Done!

**Comment by bryankaraffa** [comment]
Closing and re-opening to trigger Danger test.. looks like there may be some style issues in the Markdown docs ?

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# Turbonomic Allocate Virtual Machine Recommendations AWS
+
+## What it does
+
+The Turbonomic Allocate Virtual Machines Recommendations AWS policy utilizes Turbonomic [Credentials](https://turbonomic.com/api/v3/markets/{market_uuid}/actions) endpoint to provide AWS VM allocation recommendations for Reserved Instance coverage. From these recommendations we provide monthly savings estimates based on Turbonomic per hour costs
```
Comment (bryankaraffa) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md:5]:
```suggestion
The Turbonomic Allocate Virtual Machines Recommendations AWS policy utilizes Turbonomic Credentials (https://turbonomic.com/api/v3/markets/{market_uuid}/actions) endpoint to provide AWS VM allocation recommendations for Reserved Instance coverage. From these recommendations we provide monthly savings estimates based on Turbonomic per hour costs
```

There is no reason to hyperlink to a URL if the URL is not valid.  This is why some of the Danger errors are being raised.

This seems like a good place to mention the name of the Turbonomic API that is being used and link to the docs for the Turbonomic API (which I don't see anywhere?)

Same for all other occurrences where you have `[Credentials](https://xxxx.turbonomic.com/api/.....)`

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+        "Cookie": auth_cookie
+      }
+    }
+EOS
+end
+
+## verified that discovered is the only one we need
+script "js_get_business_units", type: "javascript" do
+  result "request", "param_turbonomic_host"
+  parameters "auth_cookie"
```
Comment (bryankaraffa) [inline cost/turbonomics/allocate_virtual_machines_recommendations/aws/turbonomics_allocate_virtual_machines.pt:155]: Some of the Danger errors are Policy Template syntax / JS errors that would get flagged when trying to upload or apply the policy templates as-is.

Can you include links to the running Applied Policies when you get them fixed please ?

File: cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md (hunk 7 above)
Comment (midir99) [reply cost/turbonomics/allocate_virtual_machines_recommendations/aws/README.md]: @bryankaraffa - Done my friend!

**Review by bryankaraffa — APPROVED** [review]
Thanks for the quick turnaround on changes!  Go ahead and merge when you're ready

**Pull request merged**